lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/NetworkSpec.hs

```diff
@@ -0,0 +1,384 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Cardano.Wallet.Jormungandr.NetworkSpec
+    ( spec
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Jormungandr.Network
+    ( UnstableBlocks (..), updateUnstableBlocks )
+import Cardano.Wallet.Primitive.Types
+    ( BlockHeader (..), Hash (..), SlotId (..) )
+import Control.Monad
+    ( (>=>) )
+import Control.Monad.Trans.Class
+    ( lift )
+import Control.Monad.Trans.Writer
+    ( execWriterT, tell )
+import Data.Coerce
+    ( coerce )
+import Data.Either
+    ( isLeft )
+import Data.List
+    ( find, findIndex, foldl', isPrefixOf )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Quantity
+    ( Quantity (..) )
+import Numeric.Natural
+    ( Natural )
+import Safe
+    ( headMay, initMay, lastMay )
+import Test.Hspec
+    ( Spec, describe, it )
+import Test.QuickCheck
+    ( Arbitrary (..)
+    , Property
+    , choose
+    , counterexample
+    , elements
+    , frequency
+    , label
+    , property
+    , vectorOf
+    , withMaxSuccess
+    , (===)
+    )
+
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.Foldable as F
+import qualified Data.Sequence as Seq
+import qualified Data.Set as Set
+
+{-------------------------------------------------------------------------------
+                                      Spec
+-------------------------------------------------------------------------------}
+
+spec :: Spec
+spec = do
+    describe "Unstable block headers" $ do
+        it "Are updated by fetching blocks" $
+            withMaxSuccess 1000 $
+            property prop_unstableBlockHeaders
+        it "Does not fetch block headers which we already have" $
+            withMaxSuccess 1000 $
+            property prop_updateUnstableBlocksIsEfficient
+        it "Handles failure of node" $
+            withMaxSuccess 1000 $
+            property prop_updateUnstableBlocksFailure
+
+{-------------------------------------------------------------------------------
+                        Unstable Block Headers Property
+-------------------------------------------------------------------------------}
+
+data TestCase = TestCase
+    { k :: Quantity "block" Natural
+    , nodeChain :: [BlockHeader]
+    , localChain :: [BlockHeader]
+    } deriving (Show, Eq)
+
+prop_unstableBlockHeaders :: TestCase -> Property
+prop_unstableBlockHeaders TestCase{..} =
+    counterexample ce prop
+  where
+    prop
+        | hasTip nodeChain = label lbl (bs' === Just expected)
+        | otherwise = label "node has no chain" (bs' === Nothing)
```

Comment:
I'd be interesting to count the number of tests cases where the node chain is actually empty and control that with a `checkCoverage`. I am not sure this case is particularly interesting so it would be nice if within the 1000 cases, we aren't testing 300 times with an empty list :sweat_smile: